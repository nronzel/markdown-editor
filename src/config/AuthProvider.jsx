import React, { useContext } from "react";

const AuthContext = React.createContext();

export const AuthProvider = ({ children, value }) => {
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export const useAuthValue = () => {
  return useContext(AuthContext);
};
