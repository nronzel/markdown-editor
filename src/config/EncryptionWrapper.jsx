import React, { useState, createContext, useContext } from "react";

export const EncryptionPasswordContext = createContext();

export const EncryptionPasswordProvider = ({ children }) => {
  const [encryptionKey, setEncryptionKey] = useState(null);
  const [isCorrectPassword, setIsCorrectPassword] = useState(false);

  async function deriveKey(password, salt) {
    const encoder = new TextEncoder();
    const passwordBuffer = encoder.encode(password);
    const saltBuffer = new Uint8Array(
      salt.match(/.{1,2}/g).map((byte) => parseInt(byte, 16))
    );
    const subtle = window.crypto.subtle;
    const key = await subtle.importKey("raw", passwordBuffer, "PBKDF2", false, [
      "deriveKey",
    ]);
    const derivedKey = await subtle.deriveKey(
      {
        name: "PBKDF2",
        salt: saltBuffer,
        iterations: 1000,
        hash: "SHA-256",
      },
      key,
      { name: "AES-GCM", length: 256 },
      false,
      ["encrypt", "decrypt"]
    );
    return derivedKey;
  }

  async function handleEncryptionKeyEntered(password, salt) {
    const derivedKey = await deriveKey(password, salt);
    setEncryptionKey(derivedKey);
    setIsCorrectPassword(true);
  }

  return (
    <EncryptionPasswordContext.Provider
      value={{
        encryptionKey,
        handleEncryptionKeyEntered,
        isCorrectPassword,
        setIsCorrectPassword,
      }}
    >
      {children}
    </EncryptionPasswordContext.Provider>
  );
};

export const useEncryptionPasswordValue = () =>
  useContext(EncryptionPasswordContext);
