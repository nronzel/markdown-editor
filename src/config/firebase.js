import { initializeApp } from "firebase/app";
import { getAuth, GoogleAuthProvider } from "firebase/auth";

const firebaseConfig = {
  apiKey: "AIzaSyD7YlbA50Vl5Qi8Qmd-LpRxwijmn7f8z2o",
  authDomain: "markdown-editor-48193.firebaseapp.com",
  projectId: "markdown-editor-48193",
  storageBucket: "markdown-editor-48193.appspot.com",
  messagingSenderId: "823260975141",
  appId: "1:823260975141:web:64d583352c28aebdc5afcc"
};

const app = initializeApp(firebaseConfig);
export const auth = getAuth(app);
export const googleProvider = new GoogleAuthProvider();
export default app
