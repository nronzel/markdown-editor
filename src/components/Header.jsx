import {
  Text,
  Button,
  Flex,
  HStack,
  Heading,
  Tooltip,
  useColorModeValue,
  useToast,
} from "@chakra-ui/react";
import React from "react";
import ThemeSwitcher from "../components/ThemeSwitcher";
import { auth } from "../config/firebase";
import { signOut } from "firebase/auth";
import { NavLink, useNavigate } from "react-router-dom";
import { faArrowRightFromBracket } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useAuthValue } from "../config/AuthProvider";

const Header = () => {
  const navigate = useNavigate();
  const { currentUser } = useAuthValue();
  const toast = useToast();

  const logOut = async () => {
    try {
      await signOut(auth);
      toast({
        title: "Logged Out Succesfully",
        status: "success",
        duration: 3000,
        isClosable: true,
      });
      navigate("/");
    } catch (err) {
      toast({
        title: `Error - ${err.message}`,
        status: "error",
        duraction: "3000",
        isClosable: true,
      });
    }
  };

  return (
    <Flex
      w="100%"
      h="70px"
      justifyContent="center"
      borderBottom="1px solid"
      borderColor={useColorModeValue("gray.300", "gray.600")}
    >
      <Flex w="80%" h="100%" justifyContent="space-between" alignItems="center">
        <Heading>
          <NavLink to="/">Markdown Editor</NavLink>
        </Heading>
        <HStack gap={5}>
          {currentUser?.isAnonymous ? (
            <Text>Anonymous</Text>
          ) : currentUser?.email ? (
            <Text>{currentUser.email}</Text>
          ) : null}
          <Flex gap={2}>
            <ThemeSwitcher />
            {currentUser ? (
              <Tooltip label="logout" borderRadius={15}>
                <Button size="sm" onClick={logOut} data-testid="logout-button">
                  <FontAwesomeIcon icon={faArrowRightFromBracket} />
                </Button>
              </Tooltip>
            ) : (
              ""
            )}
          </Flex>
        </HStack>
      </Flex>
    </Flex>
  );
};

export default Header;
