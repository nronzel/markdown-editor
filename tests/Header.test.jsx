import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Header from "../src/components/Header";
import { AuthProvider } from "../src/config/AuthProvider";
import { ChakraProvider } from "@chakra-ui/react";
import { BrowserRouter } from "react-router-dom";

function customRender(ui, { currentUser, ...options }) {
  const Wrapper = ({ children }) => (
    <AuthProvider initialUser={currentUser}>
      <ChakraProvider>
        <BrowserRouter>{children}</BrowserRouter>
      </ChakraProvider>
    </AuthProvider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
}

describe("Header", () => {
  test("renders the header with the logo and theme switcher", async () => {
    customRender(<Header />, { currentUser: null });

    expect(screen.getByText(/markdown editor/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/theme/i)).toBeInTheDocument();
  });

  test("shows 'Anonymous' on anon sign-in", async () => {
    const currentUser = { isAnonymous: true };
    customRender(<Header />, { currentUser });

    await waitFor(() => {
      expect(screen.getByText(/anonymous/i)).toBeInTheDocument();
    });
  });

  test("no logout button when not signed in", () => {
    customRender(<Header />, { currentUser: null });

    const logoutButton = screen.queryByTestId("logout-button");

    expect(logoutButton).not.toBeInTheDocument();
  });

  test("shows the email when user is logged in", async () => {
    const currentUser = { isAnonymous: false, email: "test@example.com" };
    customRender(<Header />, { currentUser });

    await waitFor(() => {
      expect(screen.getByText(/test@example.com/i)).toBeInTheDocument();
    });
  });

  test("handles logout", async () => {
    const currentUser = { isAnonymous: false, email: "test@example.com" };
    customRender(<Header />, { currentUser });

    const logoutButton = screen.getByTestId("logout-button");
    fireEvent.click(logoutButton);
    await waitFor(() => {
      expect(
        screen.queryByText(/logged out successfully/i)
      ).toBeInTheDocument();
    });
  });

  test("matches snapshot", async () => {
    const currentUser = { isAnonymous: false, email: "test@example.com" };
    const { container } = customRender(<Header />, { currentUser });

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
