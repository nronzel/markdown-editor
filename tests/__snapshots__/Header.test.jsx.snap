// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > matches snapshot 1`] = `
<div>
  <div
    class="css-e98jwe"
  >
    <div
      class="css-irlolz"
    >
      <h2
        class="chakra-heading css-1dklj6k"
      >
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          Markdown Editor
        </a>
      </h2>
      <div
        class="chakra-stack css-1t1mlqc"
      >
        <p
          class="chakra-text css-0"
        >
          test@example.com
        </p>
        <div
          class="css-1vakbk4"
        >
          <button
            aria-label="toggle theme"
            class="chakra-button css-1i07iq0"
            type="button"
          >
            <svg
              class="chakra-icon css-in1cf3"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M21.4,13.7C20.6,13.9,19.8,14,19,14c-5,0-9-4-9-9c0-0.8,0.1-1.6,0.3-2.4c0.1-0.3,0-0.7-0.3-1 c-0.3-0.3-0.6-0.4-1-0.3C4.3,2.7,1,7.1,1,12c0,6.1,4.9,11,11,11c4.9,0,9.3-3.3,10.6-8.1c0.1-0.3,0-0.7-0.3-1 C22.1,13.7,21.7,13.6,21.4,13.7z"
                fill="currentColor"
              />
            </svg>
          </button>
          <button
            class="chakra-button css-1jo2po0"
            data-testid="logout-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-right-from-bracket "
              data-icon="arrow-right-from-bracket"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M502.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-128-128c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L402.7 224 192 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l210.7 0-73.4 73.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l128-128zM160 96c17.7 0 32-14.3 32-32s-14.3-32-32-32L96 32C43 32 0 75 0 128L0 384c0 53 43 96 96 96l64 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-64 0c-17.7 0-32-14.3-32-32l0-256c0-17.7 14.3-32 32-32l64 0z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
